tools/benchmark/cmd/watch_latency.go

```diff
@@ -57,54 +63,90 @@ func init() {
 func watchLatencyFunc(_ *cobra.Command, _ []string) {
 	key := string(mustRandBytes(watchLKeySize))
 	value := string(mustRandBytes(watchLValueSize))
-
-	clients := mustCreateClients(totalClients, totalConns)
+	wchs := setupWatchChannels(key)
 	putClient := mustCreateConn()
 
-	wchs := make([]clientv3.WatchChan, len(clients))
-	for i := range wchs {
-		wchs[i] = clients[i].Watch(context.TODO(), key)
-	}
-
-	bar = pb.New(watchLTotal)
+	bar = pb.New(watchLPutTotal * len(wchs))
 	bar.Start()
 
 	limiter := rate.NewLimiter(rate.Limit(watchLPutRate), watchLPutRate)
-	r := newReport()
-	rc := r.Run()
 
-	for i := 0; i < watchLTotal; i++ {
+	putTimes := make([]time.Time, watchLPutTotal)
+	eventTimes := make([][]time.Time, len(wchs))
+
+	for i, wch := range wchs {
+		wch := wch
+		i := i
+		eventTimes[i] = make([]time.Time, watchLPutTotal)
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			eventCount := 0
+			for eventCount < watchLPutTotal {
+				resp := <-wch
+				for range resp.Events {
+					eventTimes[i][eventCount] = time.Now()
+					eventCount++
+					bar.Increment()
+				}
+			}
+		}()
+	}
+
+	putReport := newReport()
+	putReportResults := putReport.Run()
+	watchReport := newReport()
+	watchReportResults := watchReport.Run()
+	for i := 0; i < watchLPutTotal; i++ {
 		// limit key put as per reqRate
 		if err := limiter.Wait(context.TODO()); err != nil {
 			break
 		}
-
-		var st time.Time
-		var wg sync.WaitGroup
-		wg.Add(len(clients))
-		barrierc := make(chan struct{})
-		for _, wch := range wchs {
-			ch := wch
-			go func() {
-				<-barrierc
-				<-ch
-				r.Results() <- report.Result{Start: st, End: time.Now()}
-				wg.Done()
-			}()
-		}
-
+		start := time.Now()
 		if _, err := putClient.Put(context.TODO(), key, value); err != nil {
 			fmt.Fprintf(os.Stderr, "Failed to Put for watch latency benchmark: %v\n", err)
 			os.Exit(1)
 		}
-
-		st = time.Now()
-		close(barrierc)
-		wg.Wait()
-		bar.Increment()
+		end := time.Now()
+		putReport.Results() <- report.Result{Start: start, End: end}
+		putTimes[i] = end
 	}
-
-	close(r.Results())
+	wg.Wait()
+	close(putReport.Results())
 	bar.Finish()
-	fmt.Printf("%s", <-rc)
+	fmt.Printf("\nPut summary:\n%s", <-putReportResults)
+
+	for i := 0; i < len(wchs); i++ {
+		for j := 0; j < watchLPutTotal; j++ {
+			start := putTimes[j]
+			end := eventTimes[i][j]
+			if end.Before(start) {
+				start = end
+			}
+			watchReport.Results() <- report.Result{Start: start, End: end}
```

Comment:
Fragile? Puts are done sequentially and etcd guarantees that responses are sequential to writes. Not sure I would call it fragile or at least up to the etcd correctness, whoever I don't think we should care about this in benchmarks.

Reply:
The "fragile" isn't about the correctness, it's about the readability, it's about the code itself. There may be network jitter, the put may fail. Currently it looks OK because the tool just exits directly when put fails. I think we shouldn't exit in such case, we are running benchmark test, why do we care about some occasional failures due to env problems (e.g. network jitter)? Some time later, other contributors may even update the implementation not exiting in such case.

Also depending on index (correlate the times by an index) reduces the readability, also review burden. We need super carefulness to ensure there is no +1/-1 problem.

Anyway we should make the implementation more robustness & less error prone by itself.

https://github.com/etcd-io/etcd/blob/82925537e0706855436ad6944e2e543984315100/tools/benchmark/cmd/watch_latency.go#L96-L99

Reply:
I won't insist on this for now, because it works for now as mentioned above "`Currently it looks OK because the tool just exits directly when put fails.`". 

Please let me know if you want to keep it as it's or plan to change it based on the proposed above.

Reply:
Don't plan to change it for now. I think it could be improved, however I'm not as concerned with making benchmark code super readable, as it's not often changed nor critical for etcd functionality (at least until we start to really depend in automatic performance regression detection by @jmhbnz). I think we could just merge it to make sure the improvements are available for everyone to use.